singularity/local.go

```diff
@@ -0,0 +1,163 @@
+package singularity
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/data-preservation-programs/singularity/database"
+	"github.com/data-preservation-programs/singularity/datasource"
+	"github.com/data-preservation-programs/singularity/model"
+	"github.com/data-preservation-programs/singularity/service/datasetworker"
+	fs2 "github.com/rclone/rclone/fs"
+	"gorm.io/gorm"
+)
+
+var (
+	ErrItemNotFound   = errors.New("no item is found with the given id")
+	ErrSourceNotFound = errors.New("no source is found with the given id")
+	ErrNotInitialized = errors.New("singularity source not initialized")
+)
+
+const MOTION_DATASET_NAME = "MOTION_DATASET"
+const PIECE_SIZE = 1 << 36
+const MAX_CAR_SIZE = (1 << 29) * 127
+
+type LocalSingularity struct {
+	db                        *gorm.DB
+	datasourceHandlerResolver datasource.HandlerResolver
+	initialized               bool
+	sourceID                  uint32
+}
+
+func OpenSQLiteSingularity(dbFile string) (*LocalSingularity, error) {
+	db, err := database.Open("sqlite:"+dbFile, &gorm.Config{})
+	if err != nil {
+		return nil, err
+	}
+	return NewLocalSingularity(db), nil
+}
+
+func OpenMemorySingularity() (*LocalSingularity, error) {
+	db, err := database.OpenInMemory()
+	if err != nil {
+		return nil, err
+	}
+	return NewLocalSingularity(db), nil
+}
+
+func NewLocalSingularity(db *gorm.DB) *LocalSingularity {
+	return &LocalSingularity{
+		db:                        db,
+		datasourceHandlerResolver: &datasource.DefaultHandlerResolver{},
+	}
+}
+
+func (l *LocalSingularity) Start(ctx context.Context, dir string) error {
+	// migrate the db
+	if err := model.AutoMigrate(l.db); err != nil {
+		return err
+	}
+
+	// insure the creation of the default dataset, source, and root directory
+	return database.DoRetry(func() error {
+		return l.db.Transaction(func(db *gorm.DB) error {
+			var dataset model.Dataset
+			// insure dataset created
+			err := db.
+				Where(model.Dataset{Name: MOTION_DATASET_NAME}).
+				Attrs(model.Dataset{
+					MaxSize:   MAX_CAR_SIZE,
+					PieceSize: PIECE_SIZE,
+				}).FirstOrCreate(&dataset).Error
+
+			if err != nil {
+				return err
+			}
+			var source model.Source
+			err = db.
+				Where(model.Source{
+					DatasetID: dataset.ID,
+					Type:      "local",
+					Path:      dir,
+				}).Attrs(model.Source{
+				Metadata:            model.Metadata(map[string]string{}),
+				ScanIntervalSeconds: 0,
+				ScanningState:       model.Ready,
```

Comment:
If it is ready, then the first time dataset worker is run, it will start scanning this directory.
If you make it completed, then the directory will never be scanned and the only way to add item is to use pushItem